---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---

This Notebook is to compile cellranger summary files. Calculate numbers from percentages for each sample mapped to all the genome versions.
## Cellranger Summary Variables and Functions
```{r}
library(dplyr)
'%!in%' <- function(x,y)!('%in%'(x,y))
setClass("nWc")
setAs("character", "nWc", function(from) as.numeric(gsub(",", "", from) ))
setClass("nWp")
setAs("character", "nWp", function(from) as.numeric(gsub("%", "", from) ) )
countConfigPath="/data/Mullen_1/Raza/scRNA/Analysis/Configs/countConfig.txt" # to load summary files
countConfigTable=read.table(countConfigPath,sep="\t",header = T)
IdsForPlottingDf<-read.table("/data/Mullen_1/Raza/scRNA/Analysis/Configs/IDs_For_Plotting.txt",sep="\t",header = T)
countConfigTable=countConfigTable[which(countConfigTable$baseDir!="GSE156337"),] # remove this dataset. 
referenceNamesForID=c("TRIMMED","UNTRIMMED","GENCODE")
resultsDir="/data/Mullen_1/Raza/scRNA/Results/"
mappingDir=paste0(resultsDir,"mapping_1/")
cellRangerSummaryDir=paste0(resultsDir,"DOWNSTREAM/CELLRANGER_SUMMARY/")
if(!dir.exists(cellRangerSummaryDir)){dir.create(cellRangerSummaryDir,mode="0777",recursive = T)}

```

## Cellranger Summary Percentages
```{r}
cellRangerSummaryDF=data.frame()
for(referenceName in referenceNamesForID){
  for(i in 1:nrow(countConfigTable)){
    datasetID=countConfigTable[i,"baseDir"]
    baseResDir=paste0(countConfigTable[i,"baseDir"],"/")
    id=countConfigTable[i,"id"]
    summaryFile=paste0(mappingDir,baseResDir,referenceName,"/",id,"/",id,"/outs/metrics_summary.csv")
    if(file.exists(summaryFile)){
      x=read.table(summaryFile,sep=",",header = T,colClasses=c('nWc','nWc','nWc','nWc','nWp','nWp','nWp','nWp','nWp','nWp','nWp','nWp','nWp','nWp','nWp','nWp','nWp','nWc','nWc'))
      x[1,"DatasetId"]=datasetID
      x[1,"SampleId"]=id
      x[1,"Reference"]=referenceName
      rownames(x)=paste0(id,"__",referenceName)
      cellRangerSummaryDF=rbind(cellRangerSummaryDF,x)
    }
  }
}
columnOrder=c(which(colnames(cellRangerSummaryDF) %in% c("Reference","DatasetId","SampleId")),which(colnames(cellRangerSummaryDF) %!in% c("Reference","DatasetId","SampleId")))
cellRangerSummaryDF <- cellRangerSummaryDF[, columnOrder]
write.table(cellRangerSummaryDF,file=paste0(cellRangerSummaryDir,"/cellRangerSummary_percentages.txt"), col.names  = NA,sep="\t",quote = F)

```

## Cellranger Summary Percentages with numbers
```{r}
calculateNumberForPercentages<-c("Reads.Mapped.Confidently.to.Genome","Reads.Mapped.Confidently.to.Intergenic.Regions","Reads.Mapped.Confidently.to.Intronic.Regions","Reads.Mapped.Confidently.to.Exonic.Regions","Reads.Mapped.Confidently.to.Transcriptome","Reads.Mapped.Antisense.to.Gene")
numbersFromPercentagesDF<-(cellRangerSummaryDF[,calculateNumberForPercentages]/100)*cellRangerSummaryDF[,"Number.of.Reads"]
numbersFromPercentagesDF=round(numbersFromPercentagesDF,digits = 2)
colnames(numbersFromPercentagesDF)<-paste0("Number_of_",colnames(numbersFromPercentagesDF))
cellRangerSummaryDF<-cbind(cellRangerSummaryDF,numbersFromPercentagesDF)
write.table(numbersFromPercentagesDF,file=paste0(cellRangerSummaryDir,"/cellRangerSummary_numbers.txt"), col.names  = NA,sep="\t",quote = F)
write.table(cellRangerSummaryDF,file=paste0(cellRangerSummaryDir,"/cellRangerSummary.txt"), col.names  = NA,sep="\t",quote = F)

```

## Compare within each reference genome (GENCODE, TRIMMED, UNTRIMMED)
```{r}
library(dplyr)
totalReads=0
if(sum(cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="GENCODE"),"Number.of.Reads"])==sum(cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="TRIMMED"),"Number.of.Reads"]) &
   sum(cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="GENCODE"),"Number.of.Reads"])==sum(cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="UNTRIMMED"),"Number.of.Reads"])){
  totalReads=sum(cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="GENCODE"),"Number.of.Reads"])}

print(paste0("totalReads ",totalReads))
g=cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="GENCODE"),colnames(numbersFromPercentagesDF)]
t=cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="TRIMMED"),colnames(numbersFromPercentagesDF)]
u=cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="UNTRIMMED"),colnames(numbersFromPercentagesDF)]
if(nrow(cellRangerSummaryDF)==(nrow(g)+nrow(t)+nrow(u))){print("Correct split")}

# SUM
gSum = round(summarise(g,across(where(is.numeric), sum),across(where(is.character), ~0)),digits = 2)
tSum = round(summarise(t,across(where(is.numeric), sum),across(where(is.character), ~0)),digits = 2)
uSum = round(summarise(u,across(where(is.numeric), sum),across(where(is.character), ~0)),digits = 2)

## rownames 
rownames(gSum)<-"Sum"
rownames(tSum)<-"Sum"
rownames(uSum)<-"Sum"

# bind total and mean
g<-rbind(g,gSum)
t<-rbind(t,tSum)
u<-rbind(u,uSum)

write.table(g,file=paste0(cellRangerSummaryDir,"/GENCODE.txt"), col.names  = NA,sep="\t",quote = F)
write.table(t,file=paste0(cellRangerSummaryDir,"/TRIMMED_SINGLETROME.txt"), col.names  = NA,sep="\t",quote = F)
write.table(u,file=paste0(cellRangerSummaryDir,"/UNTRIMMED_SINGLETROME.txt"), col.names  = NA,sep="\t",quote = F)

## Difference 
gTot=g["Sum",]-t["Sum",]
gTou=g["Sum",]-u["Sum",]
tTou=t["Sum",]-u["Sum",]
## Difference To percentages
gTotPercentages=(gTot/totalReads)*100
gTouPercentages=(gTou/totalReads)*100
tTouPercentages=(tTou/totalReads)*100

#
comparisonsDf<-t(rbind("GENCODE-TRIMMED"=gTot,"GENCODE-TRIMMED (%)"=gTotPercentages,"GENCODE-UNTRIMMED"=gTou,"GENCODE-UNTRIMMED (%)"=gTouPercentages,"TRIMMED-UNTRIMMED"=tTou,"TRIMMED-UNTRIMMED (%)"=tTouPercentages))
comparisonsDf=round(comparisonsDf,digits = 2)
write.table(comparisonsDf,file=paste0(cellRangerSummaryDir,"/Difference_in_genome_mapping.txt"), col.names  = NA,sep="\t",quote = F)

# to millions, only for writing purposes
df1=comparisonsDf
df1[,c("GENCODE-TRIMMED", "GENCODE-UNTRIMMED","TRIMMED-UNTRIMMED")]<- round(df1[,c("GENCODE-TRIMMED", "GENCODE-UNTRIMMED","TRIMMED-UNTRIMMED")]/1e6,digits = 2)
write.table(df1,file=paste0(cellRangerSummaryDir,"/Difference_in_genome_mapping_to_millions_for_writing.txt"), col.names  = NA,sep="\t",quote = F)

### Just to see the total genes detected
genCodeNumberOfDetectedGenes=cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="GENCODE"),"Total.Genes.Detected",]
trimmedNumberOfDetectedGenes=cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="TRIMMED"),"Total.Genes.Detected",]
unTrimmedNumberOfDetectedGenes=cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="UNTRIMMED"),"Total.Genes.Detected",]
mean(genCodeNumberOfDetectedGenes)-mean(unTrimmedNumberOfDetectedGenes)

```
#JUST to cross check the above calculations. It is correct. I just commented it to not re-run it every time.
```{r}
g=cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="GENCODE"),]
u=cellRangerSummaryDF[which(cellRangerSummaryDF$Reference=="UNTRIMMED"),]
# Calculate average of percentages https://www.indeed.com/career-advice/career-development/how-to-calculate-average-percentage#:~:text=Calculate%20the%20percentage%20average,to%20get%20the%20average%20percentage.
a=((sum((g[,"Reads.Mapped.Confidently.to.Exonic.Regions"]/100)*(g[,"Number.of.Reads"])))/sum(g[,"Number.of.Reads"]))*100
b=((sum((u[,"Reads.Mapped.Confidently.to.Exonic.Regions"]/100)*(u[,"Number.of.Reads"])))/sum(u[,"Number.of.Reads"]))*100
c=((a-b)/100)*(sum(g[,"Number.of.Reads"]))
print(paste(a, b, a-b,c, sep="    "))
```

## Plots
```{r}
source("CustomTheme.R")
library("ggpubr")
parameters=c("Total.Genes.Detected","Reads.Mapped.Confidently.to.Exonic.Regions","Reads.Mapped.Confidently.to.Genome","Reads.Mapped.Confidently.to.Transcriptome","Reads.Mapped.Confidently.to.Intronic.Regions","Reads.Mapped.Confidently.to.Intergenic.Regions")
parameterLabels=c("Total.Genes.Detected","Unique.Exonic.mapping","Unique.Genome.mapping","Unique.Transcriptome.mapping","Unique.Intronic.mapping","Unique.Intergenic.mapping")
```
# MAIN FIGURES
```{r}
ds="ALL"
#tempDF=cellRangerSummaryDF[which(cellRangerSummaryDF$DatasetId %in% c("GSE115469","pbmc_10k_v3")),]
tempDF=cellRangerSummaryDF
plotDir<-paste0(cellRangerSummaryDir,"/",ds,"/")
if(!dir.exists(plotDir)){dir.create(plotDir,mode="0777",recursive = T)}
tempDF <- merge(tempDF, IdsForPlottingDf, by.x=c("SampleId"), by.y=c("SampleId"), all=F)
tempDF[which(tempDF$Reference=="TRIMMED"),"Reference"]="TLGA"
tempDF[which(tempDF$Reference=="UNTRIMMED"),"Reference"]="ULGA"
allPlots<-list()
for(i in 1:length(parameters)){
  parameter=parameters[i]
  if(parameter=="Total.Genes.Detected"){ ylabel="Number of Genes"}else{ ylabel="Percentage (%)"}
  toPlotDF=tempDF[,c(parameter,"DatasetId","Reference","newSampleId")]
  colnames(toPlotDF)[1]="value"
  title=tolower(gsub("\\."," ",parameterLabels[i]))
  title=paste(toupper(substring(title, 1,1)),substring(title,2) ,sep="", collapse=" ")
  toPlotDF$Reference<-as.factor(as.character(toPlotDF$Reference))
 
  pdf(paste0(plotDir,"/",parameter,".pdf"))
  p=ggplot(toPlotDF,aes(x=newSampleId,y=value,fill=Reference))+geom_col(width = 0.7, position = "dodge")  +ggtitle(title) + labs(y = ylabel) +theme_Publication() + scale_fill_Publication() + xlab("") +theme(axis.text.x = element_text(angle = 45, hjust=1,size = rel(1.2)),axis.text.y = element_text(size = rel(1)),axis.title = element_text(face = "bold",size = rel(1.5)),legend.key.size = unit(0.9, 'cm'),legend.text = element_text(size=12)) + scale_x_discrete(limits = IdsForPlottingDf$newSampleId) +
    scale_fill_manual(values=c("#fdb462","#7fc97f","#386cb0"))
  print(p)
  dev.off()
  allPlots[[i]]=p
}
figure <- ggarrange(allPlots[[1]],allPlots[[2]],allPlots[[3]],allPlots[[4]],allPlots[[5]],allPlots[[6]], labels = c("A.", "B.", "C.","D.","E.","F."), ncol = 1,nrow=6,font.label=list(size=20),common.legend = TRUE)
ggexport(figure, filename = paste0(plotDir,"/",ds,"_CellRangerSummary.pdf"),width = 8, height = 24)

```