---
  title: "R Notebook"
output:
  html_document:
  df_print: paged
---
  
  
## Libraries
  
```{r results='hide', message=FALSE, warning=FALSE}
require(ggplot2)
require(Seurat)
require(reshape2)
library("ggpubr")
library(car) # for scatterplot
#source("FUNCTIONS.R")
source("CustomTheme.R")
```
## directories and vairables
```{r}
resultsDir="Results/"
clusteringDir=paste0(resultsDir,"DOWNSTREAM/CLUSTERING/")
filteredTranscriptsDir=paste0(resultsDir,"DOWNSTREAM/FILTERED_TRANSCRIPTS/")
if(!dir.exists(filteredTranscriptsDir)){dir.create(filteredTranscriptsDir,mode="0777",recursive = T)}
analysisConfigPath="Analysis/Configs/analysisConfig.txt" 
analysisConfigTable=read.table(analysisConfigPath,sep="\t",header = T)
# antiSenseOverlapGenecodeBedFile
antiSenseOverlapGenecodeBedFile="Genomes/OverlapCleansing/antiStrand_LnRNA_ExonOverlap_ProteinCodingExons.bed"
antiSenseOverlapGenecodeBed = read.table(antiSenseOverlapGenecodeBedFile, sep='\t',header = F)
colnames(antiSenseOverlapGenecodeBed) = c("queryChr","queryStart","queryEnd","queryGeneId","queryScore","queryStrand","refChr","refStart","refEnd","refGeneId","refScore","refStrand","overlapLength")
# To obtain the gene names for the gene ids because cellrange use gene name if it exits

transcriptlengthTable<-read.table("Genomes/GTFDetails/transcriptAndGene.txt",sep="\t",header = T,row.names = 1)
antsenseAllGeneIdsAndNames<-transcriptlengthTable[which(transcriptlengthTable$gene_id %in% antiSenseOverlapGenecodeBed$queryGeneId),"geneName"]
antsenseAllGeneIdsAndNames<-unique(antsenseAllGeneIdsAndNames[which(!is.na(antsenseAllGeneIdsAndNames))])
antsenseAllGeneIdsAndNames=c(antsenseAllGeneIdsAndNames,paste0(antsenseAllGeneIdsAndNames,".1"),paste0(antsenseAllGeneIdsAndNames,".2"),paste0(antsenseAllGeneIdsAndNames,".3"),paste0(antsenseAllGeneIdsAndNames,".4"))
antsenseAllGeneIdsAndNames<-c(antsenseAllGeneIdsAndNames,antiSenseOverlapGenecodeBed$queryGeneId)

```

## DOWNSTREAM ANALYSIS
# COMPARE trimmed and untrimmed expression
# COMPARE PCG and lncRNA expression
# COMPARE lncRNA expression across different categories
```{r results='hide', message=FALSE, warning=FALSE}
ncountRNACombinedAllPlots<-list()
ncountRNAPerCellTypeAllPlots<-list()
title<-c("PBMCs","Liver","Liver")
trimiUntrimFilterDir=paste0(filteredTranscriptsDir,"/TRIM_UNTRIM_FILTER/")
if(!dir.exists(trimiUntrimFilterDir)){dir.create(trimiUntrimFilterDir,mode="0777",recursive = T)}
for(dsRow in 1:nrow(analysisConfigTable)){
  datasetID=analysisConfigTable[dsRow,"baseDir"]
  currentDir=paste0(trimiUntrimFilterDir,datasetID,"/")
  if(!dir.exists(currentDir)){dir.create(currentDir,mode="0777",recursive = T)}
  trimmedLncRNAObj=readRDS(paste0(clusteringDir,datasetID,"/TRIMMED/LNCRNA.RDS"))
  unTrimmedLncRNAObj=readRDS(paste0(clusteringDir,datasetID,"/UNTRIMMED/LNCRNA.RDS"))
  unTrimmedCombinedObj=readRDS(paste0(clusteringDir,datasetID,"/UNTRIMMED/COMBINED.RDS"))
  ############################################################################### Only restrict to antisense overlapping genes
  subTrimmed <- subset(trimmedLncRNAObj, features = unique(antsenseAllGeneIdsAndNames))
  subUn <- subset(unTrimmedLncRNAObj, features = unique(antsenseAllGeneIdsAndNames))
  ############################################################################### # Find the difference in the trimmed and unrtimmed genome versions
  onlyInUnTrimmedObj<-subset(subUn, features = unique(setdiff(rownames(subUn), rownames(subTrimmed))))
  onlyInTrimmedObj<-subset(subTrimmed, features = unique(setdiff(rownames(subTrimmed), rownames(subUn)))) # not used
  tCommonInBothObj<-subset(subTrimmed, features = unique(intersect(rownames(subUn), rownames(subTrimmed))))
  unCommonInBothObj<-subset(subUn, features = unique(intersect(rownames(subUn), rownames(subTrimmed))))
  write.table(file=paste0(currentDir,"/antisense_onlyInUnTrimmed.txt"),rownames(onlyInUnTrimmedObj),quote = F,row.names = F,col.names = F)
  write.table(file=paste0(currentDir,"/antisense_onlyInTrimmed.txt"),rownames(onlyInTrimmedObj),quote = F,row.names = F,col.names = F)
  write.table(file=paste0(currentDir,"/antisense_commonInBoth.txt"),unique(intersect(rownames(subUn), rownames(subTrimmed))),quote = F,row.names = F,col.names = F)
  ############################################################################### # stats
  row=c("Total Genes Untrimmed"=nrow(unTrimmedCombinedObj), "Total lncRNAs Untrimmed"=nrow(unTrimmedLncRNAObj), "Total lncRNAs Trimmed"= nrow(trimmedLncRNAObj), "Overlapping antisense lncRNAs Untrimmed"=nrow(subUn),"Overlapping antisense lncRNAs Trimmed"=nrow(subTrimmed), "Only In UnTrimmed"=nrow(onlyInUnTrimmedObj), "Only In Trimmed"=nrow(onlyInTrimmedObj), "Common In Both"=length(unique(intersect(rownames(subUn), rownames(subTrimmed)))), "Median unTrimmed Counts CommonInBoth"= median(unCommonInBothObj@meta.data$nCount_RNA),"Median Trimmed Counts CommonInBoth"= median(tCommonInBothObj@meta.data$nCount_RNA))
############################################################################### # Expression
  onlyInUnTrimmedGeneSum <- as.matrix(Matrix::rowSums(GetAssayData(object = onlyInUnTrimmedObj, slot = "counts")))
  onlyInUnTrimmedGeneSum= as.data.frame(onlyInUnTrimmedGeneSum[order(onlyInUnTrimmedGeneSum[,1],decreasing=T),])
  colnames(onlyInUnTrimmedGeneSum)<-"counts"
  
  commonInUntrimmed <- as.matrix(Matrix::rowSums(GetAssayData(object = unCommonInBothObj, slot = "counts")))
  commonInUntrimmed= as.data.frame(commonInUntrimmed[order(commonInUntrimmed[,1],decreasing=T),])
  colnames(commonInUntrimmed)<-"counts_commonInUntrimmed"

  commonInTRIMMED <- as.matrix(Matrix::rowSums(GetAssayData(object = tCommonInBothObj, slot = "counts")))
  commonInTRIMMED= as.data.frame(commonInTRIMMED[order(commonInTRIMMED[,1],decreasing=T),])
  colnames(commonInTRIMMED)<-"counts_commonInTRIMMED"

  commGenesExpDf<-merge(commonInUntrimmed,commonInTRIMMED,by="row.names")
  rownames(commGenesExpDf)<-commGenesExpDf$Row.names
  commGenesExpDf$Row.names<-NULL
  commGenesExpDf$ratio=commGenesExpDf$counts_commonInUntrimmed/commGenesExpDf$counts_commonInTRIMMED
  commGenesExpDf=commGenesExpDf[order(-commGenesExpDf$ratio),]
  
  write.table(file=paste0(currentDir,"/onlyInUnTrimmedCounts.txt"),onlyInUnTrimmedGeneSum,quote = F,col.names = NA)
  write.table(file=paste0(currentDir,"/commGenesCounts.txt"),commGenesExpDf,quote = F,col.names = NA)
  ############################################################################### # PLOT
  reOrderVlnPlot <- c("Antisense only","ULGA","TLGA")
  onlyInUnTrimmedObj@meta.data$type=reOrderVlnPlot[1]
  unCommonInBothObj@meta.data$type=reOrderVlnPlot[2]
  tCommonInBothObj@meta.data$type=reOrderVlnPlot[3]
  combined <- merge(onlyInUnTrimmedObj, y = c(unCommonInBothObj,tCommonInBothObj), add.cell.ids = c("onlyInUnTrimmedObj", "unCommonInBothObj","tCommonInBothObj"),
                    project = "combined")
  combined@meta.data$type <- factor(x = combined@meta.data$type, levels = reOrderVlnPlot)

nCount_RNA_Combined<-VlnPlot(combined, features = c("nCount_RNA"), group.by = "type",log=T,pt.size = 0)+geom_boxplot(width=0.1,fill="white") + ggtitle(paste0(title[dsRow])) +
  theme_Publication() + scale_fill_Publication() + ylab("log (Number of reads)") +xlab("") + theme(axis.text.x = element_blank(),axis.text.y = element_text(size = 12),axis.title = element_text(face = "bold",size = 12), plot.title = element_text(size = 12, face = "bold"),legend.key.size = unit(0.6, 'cm'),legend.text = element_text(size=12))

nCount_RNA_PerCellType<-VlnPlot(combined, features = c("nCount_RNA"), group.by = "celltype",split.by = "type", log=T,pt.size = 0) +
  ggtitle(paste0(title[dsRow]," per cell type")) +
  theme_Publication() + scale_fill_Publication() + xlab("") + ylab("log (Number of reads)") +theme_Publication() + scale_fill_Publication() + xlab("") + ylab("log (Number of reads)") +theme(plot.title = element_text(size = 12, face = "bold"), axis.text.x = element_text(angle = 45, hjust=1,size = 12),axis.text.y = element_text(size = 12),axis.title = element_text(face = "bold",size = 12),legend.key.size = unit(0.6, 'cm'),legend.text = element_text(size=12)) 

nCountRNAFigure <- ggarrange(nCount_RNA_Combined,nCount_RNA_PerCellType, labels = c("A", "B"), ncol = 2, nrow = 1,common.legend = TRUE,widths = c(0.7,2))
ggexport(nCountRNAFigure, filename = paste0(currentDir,datasetID,"_nCount_RNA.pdf"),width = 8, height = 5)

ncountRNACombinedAllPlots[[dsRow]]=nCount_RNA_Combined
ncountRNAPerCellTypeAllPlots[[dsRow]]=nCount_RNA_PerCellType
  ############################################################################### # remove genes
  unTrimmedLncRNAObj <- unTrimmedLncRNAObj[!rownames(unTrimmedLncRNAObj) %in% unique(rownames(onlyInUnTrimmedObj)),]
  unTrimmedCombinedObj <- unTrimmedCombinedObj[!rownames(unTrimmedCombinedObj) %in% unique(rownames(onlyInUnTrimmedObj)),]
  
  row=c(row,"unTrimmed Genes after deleton"=nrow(unTrimmedLncRNAObj),"Singletrome:Total Genes Untrimmed after deleton"=nrow(unTrimmedCombinedObj))
  write.table(file=paste0(currentDir,"Filter_based_on_trimmed_stats.txt"),row,sep="\t",col.names = F,quote = F)    
  saveRDS(unTrimmedLncRNAObj,file = paste0(currentDir,"unTrimmedLncRNAObj.RDS"))
  saveRDS(unTrimmedCombinedObj,file = paste0(currentDir,"Singletrome.RDS"))
}
figure <- ggarrange(ncountRNACombinedAllPlots[[1]],ncountRNAPerCellTypeAllPlots[[1]],ncountRNACombinedAllPlots[[2]],ncountRNAPerCellTypeAllPlots[[2]],ncountRNACombinedAllPlots[[3]],ncountRNAPerCellTypeAllPlots[[3]], labels = c("A.", "B.", "C.","D.","E.","F."), ncol = 2, nrow = 3,common.legend = TRUE,widths = c(0.7,2), align="hv",font.label=list(size=20),legend = "bottom")
ggexport(figure, filename = paste0(trimiUntrimFilterDir,"/","ncountRNA_AllPlots.pdf"),width = 8, height = 15)
```