{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modified from TALON: Techonology-Agnostic Long Read Analysis Pipeline. Authored by : Dana Wyman\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import sys\n",
    "import itertools\n",
    "from copy import deepcopy\n",
    "import copy\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Edge(object):\n",
    "    \"\"\"Stores information about an edge, including its location\n",
    "       and the gene/transcript(s) it belongs to.\n",
    "       Attributes:\n",
    "           identifier: Accession ID of the edge\n",
    "           gene: Accession ID of the gene that the edge belongs to\n",
    "           transcript_ids: Set of transcript accession IDs that the edge \n",
    "           belongs to\n",
    "           chromosome: Chromosome that the transcript is located on \n",
    "           (format \"chr1\")\n",
    "           start: The start position of the edge with respect to the\n",
    "           forward strand \n",
    "           end: The end position of the edge with respect to the\n",
    "           forward strand\n",
    "           strand: \"+\" if the edge is on the forward strand, and \"-\" if\n",
    "           it is on the reverse strand\n",
    " \n",
    "           length: The length of the edge\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, identifier, chromosome, start, end, strand, gene_id,\n",
    "                 transcript_id, annotations):\n",
    "        self.chromosome = str(chromosome)\n",
    "        self.gene_id = gene_id\n",
    "        self.start = int(start)\n",
    "        self.end = int(end)\n",
    "        self.strand = strand\n",
    "        self.length = abs(self.end - self.start + 1)\n",
    "        self.annotations = annotations\n",
    "\n",
    "        self.identifier = str(identifier)\n",
    "        self.transcript_ids = set()\n",
    "        if transcript_id != None:\n",
    "            self.transcript_ids.add(transcript_id)\n",
    "        self.v1 = None\n",
    "        self.v2 = None\n",
    "\n",
    "    def print_edge(self):\n",
    "        \"\"\" Prints a string representation of the edge\"\"\"\n",
    "        print(self.identifier + \": \" + self.chromosome + \":\" + \\\n",
    "              str(self.start) + \"-\" + str(self.end))\n",
    "        print(self.transcript_ids)\n",
    "        return\n",
    "    \n",
    "    def getExonLineByObject(exonObj):\n",
    "        exonLine=str(exonObj.chromosome+\"\\t\"+exonObj.annotations.get('source')+\"\\t\"+\"exon\"+\"\\t\"+str(exonObj.start)+\"\\t\"+str(exonObj.end)+\"\\t.\\t\"+exonObj.strand+\"\\t.\\t\")\n",
    "        for x in exonObj.annotations:\n",
    "            exonLine=exonLine+str(x+\" \\\"\"+exonObj.annotations[x]+\"\\\"; \")\n",
    "        return exonLine+\"\\n\"\n",
    "\n",
    "    def create_edge_from_gtf(edge_info):\n",
    "        \"\"\" Creates an edge object using information from a GTF entry\n",
    "                Args:\n",
    "                   edge_info: A list containing fields from a GTF file edge entry.\n",
    "                   Example:   \n",
    "                   ['chr1', 'HAVANA', 'exon', '11869', '12227', '.', '+', '.', \n",
    "                    'gene_id \"ENSG00000223972.5\"; transcript_id \"ENST00000456328.2\"; \n",
    "                    gene_type \"transcribed_unprocessed_pseudogene\"; \n",
    "                    gene_status \"KNOWN\"; gene_name \"DDX11L1\"; \n",
    "                    transcript_type \"processed_transcript\"; \n",
    "                    transcript_status \"KNOWN\"; transcript_name \"DDX11L1-002\"; \n",
    "                    edge_number 1; edge_id \"ENSE00002234944.1\"; level 2; \n",
    "                    tag \"basic\"; transcript_support_level \"1\"; \n",
    "                    havana_gene \"OTTHUMG00000000961.2\"; \n",
    "                    havana_transcript \"OTTHUMT00000362751.1\";'] \n",
    "        \"\"\"\n",
    "        description = edge_info[-1]\n",
    "        start = int(edge_info[3])\n",
    "        end = int(edge_info[4])\n",
    "        chromosome = edge_info[0]\n",
    "        strand = edge_info[6]\n",
    "\n",
    "        annotations = extract_edge_annotations_from_GTF(edge_info)\n",
    "        if \"exon_id\" not in annotations:\n",
    "            annotations[\"exon_id\"] = \"_\".join([chromosome, str(start), str(end), strand])\n",
    "        gene_id = annotations['gene_id']\n",
    "        transcript_id = annotations['transcript_id']\n",
    "        edge_id = \"_\".join([chromosome, str(start), str(end), strand,gene_id])\n",
    "\n",
    "        if \"gene_id\" in description:\n",
    "            gene_id = (description.split(\"gene_id \")[1]).split('\"')[1]\n",
    "        if \"transcript_id\" in description:\n",
    "            transcript_id = (description.split(\"transcript_id \")[1]).split('\"')[1]\n",
    "\n",
    "        edge = Edge(edge_id, chromosome, start, end, strand, gene_id, transcript_id,\n",
    "                    annotations)\n",
    "        return edge\n",
    "\n",
    "def extract_edge_annotations_from_GTF(tab_fields):\n",
    "    \"\"\" Extracts key-value annotations from the GTF description field\n",
    "    \"\"\"\n",
    "\n",
    "    attributes = {}\n",
    "\n",
    "    # remove trailing newline and split by semicolon\n",
    "    description = tab_fields[-1].strip('\\n')\n",
    "    description = description.split(';')\n",
    "\n",
    "    # Parse description\n",
    "    for fields in description:\n",
    "        if fields == \"\" or fields == \" \": continue\n",
    "        fields = fields.split()\n",
    "        if fields[0] == '': fields = fields[1:]\n",
    "\n",
    "        key = fields[0].replace('\"', '')\n",
    "        val = ' '.join(fields[1:]).replace('\"', '')\n",
    "        \n",
    "        attributes[key] = val\n",
    "\n",
    "    # Put in placeholders for important attributes (such as gene_id) if they\n",
    "    # are absent\n",
    "    if \"gene_id\" not in attributes:\n",
    "        attributes[\"gene_id\"] = \"NULL\"\n",
    "    if \"transcript_id\" not in attributes:\n",
    "        attributes[\"transcript_id\"] = \"NULL\"\n",
    "\n",
    "    attributes[\"source\"] = tab_fields[1]\n",
    "\n",
    "    return attributes\n",
    "\n",
    "def get_edge_from_db(vertex_info_1, vertex_info_2):\n",
    "    \"\"\" Uses information from a database edge entry to create an edge object.\n",
    "    \"\"\"\n",
    "    if vertex_info_1[\"edge_id\"] != vertex_info_2[\"edge_id\"]:\n",
    "        raise ValueError('Tried to create edge from endpoints with different IDs')\n",
    "    edge_id = vertex_info_1[\"edge_id\"]\n",
    "    chromosome = vertex_info_1['chromosome']\n",
    "    start = min(vertex_info_1['position'], vertex_info_2['position'])\n",
    "    end = max(vertex_info_1['position'], vertex_info_2['position']) \n",
    "    strand = vertex_info_1['strand']\n",
    "    gene_id = vertex_info_1['gene_id']\n",
    "\n",
    "    edge = Edge(edge_id, chromosome, start, end, strand, gene_id, None, None)\n",
    "    edge.v1 = str(vertex_info_1[\"vertex_ID\"])\n",
    "    edge.v2 = str(vertex_info_2[\"vertex_ID\"])\n",
    "    return edge\n",
    "\n",
    "def create_novel_edge(chromosome, start, end, strand, gene_id, transcript_id, counter):\n",
    "    \"\"\" Creates a novel edge with a unique identifier (obtained using\n",
    "        counter). Returns the edge object as well as the updated counter.\n",
    "    \"\"\"\n",
    "    counter[\"edges\"] += 1\n",
    "    curr_novel = counter[\"edges\"]\n",
    "    edge = Edge(curr_novel, chromosome, start, end, strand, gene_id, transcript_id,\n",
    "                None)\n",
    "    return edge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Transcript(object):\n",
    "    \"\"\"Stores information about a gene transcript, including its location\n",
    "       and constitutive exons.\n",
    "       Attributes:\n",
    "           identifier: Accession ID of transcript, i.e. an Ensembl ID. Must\n",
    "           be unique.\n",
    "           name: Human-readable name of the transcript. Does not have to be \n",
    "           unique\n",
    "           chromosome: Chromosome that the transcript is located on \n",
    "           (format \"chr1\")\n",
    "           start: The start position of the transcript with respect to the\n",
    "           forward strand \n",
    "           end: The end position of the transcript with respect to the\n",
    "           forward strand\n",
    "           strand: \"+\" if the transcript is on the forward strand, and \"-\" if\n",
    "           it is on the reverse strand\n",
    "           gene_id: unique ID of the gene that this transcript belongs to\n",
    "           exons: List of exon objects belonging to this transcript, in sorted\n",
    "           order.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, identifier, chromosome, start, end, strand, gene_id, \n",
    "                 annotations):\n",
    "\n",
    "        self.identifier = str(identifier)\n",
    "        self.gene_id = str(gene_id)\n",
    "\n",
    "        self.chromosome = str(chromosome)\n",
    "        self.start = int(start)\n",
    "        self.end = int(end)\n",
    "        self.strand = strand\n",
    "        self.n_exons = 0\n",
    "        self.exons = []\n",
    "        self.introns = []\n",
    "        self.annotations = annotations\n",
    "\n",
    "    def get_5prime_vertex(self):\n",
    "        \"\"\" Returns ID of 5' end vertex \"\"\"\n",
    "\n",
    "        if self.strand == \"+\":\n",
    "            return self.exons[0].v1\n",
    "        if self.strand == \"-\":\n",
    "            return self.exons[-1].v2\n",
    "\n",
    "    def get_3prime_vertex(self):\n",
    "        \"\"\" Returns ID of 5' end vertex \"\"\"\n",
    "\n",
    "        if self.strand == \"+\":\n",
    "            return self.exons[-1].v2\n",
    "        if self.strand == \"-\":\n",
    "            return self.exons[0].v1\n",
    "\n",
    "    def get_edge_path(self):\n",
    "        edges = self.get_all_edges()\n",
    "        if len(edges) == 0:\n",
    "            return None\n",
    "        path = [ x.identifier for x in edges]\n",
    "\n",
    "        # Must reverse the path if the transcript is on the '-' strand\n",
    "        if self.strand == \"-\":\n",
    "            path = path[::-1]\n",
    "        return \",\".join(path)\n",
    "\n",
    "    def get_all_edges(self):\n",
    "        all_edges = []\n",
    "        for i in range(0,self.n_exons):\n",
    "            all_edges.append(self.exons[i])\n",
    "            try:\n",
    "                all_edges.append(self.introns[i])\n",
    "            except:\n",
    "                pass\n",
    "            \n",
    "        return all_edges\n",
    "\n",
    "    def get_length(self):\n",
    "        \"\"\" Computes the length of the transcript by summing the lengths of\n",
    "            its exons \"\"\"\n",
    "\n",
    "        if len(self.exons) == 0:\n",
    "            raise ValueError('Cannot compute length: Transcript does not ' + \\\n",
    "                             'have any exons')\n",
    "        \n",
    "        transcript_length = 0\n",
    "        for exon in self.exons:\n",
    "            transcript_length += exon.length\n",
    "        return transcript_length\n",
    "\n",
    "    def get_exon_coords(self):\n",
    "        \"\"\" Returns a list of the exon coordinates in order \"\"\"\n",
    "        exon_coords = []\n",
    "        for exon in self.exons:\n",
    "            exon_coords.append(int(exon.start))\n",
    "            exon_coords.append(int(exon.end))\n",
    "        return exon_coords\n",
    "\n",
    "    def add_exon(self, exon):\n",
    "        \"\"\"Adds an exon object to the transcript.\"\"\"\n",
    "\n",
    "        if exon.start > exon.end:\n",
    "            raise ValueError('Exon start (' + str(exon.start) + ') ' + \\\n",
    "                'is supposed to be before the exon end (' + str(exon.end) + ')')\n",
    "\n",
    "        # Check where in the list the exon should be added\n",
    "        for i in range(0,len(self.exons)):\n",
    "            existing_exon = self.exons[i]\n",
    "            if exon.end < existing_exon.start:\n",
    "                self.exons = self.exons[0:i] + [exon] + self.exons[i:]\n",
    "                self.check_exon_validity()\n",
    "                self.n_exons += 1\n",
    "                return\n",
    "        self.exons.append(exon)\n",
    "        self.check_exon_validity()\n",
    "        self.n_exons += 1\n",
    "        return\n",
    "\n",
    "    def add_intron(self, intron):\n",
    "        \"\"\"Adds an edge object to the transcript.\"\"\"\n",
    "\n",
    "        if intron.start > intron.end:\n",
    "            raise ValueError('Intron start (' + str(intron.start) + ')' + \\\n",
    "                'is supposed to be before the intron end (' + str(intron.end) + ')')\n",
    "\n",
    "        # Check where in the list the intron should be added\n",
    "        for i in range(0,len(self.introns)):\n",
    "            existing_intron = self.introns[i]\n",
    "            if intron.end < existing_intron.start:\n",
    "                self.introns = self.introns[0:i] + [intron] + self.introns[i:]\n",
    "                return\n",
    "        self.introns.append(intron)\n",
    "        return\n",
    "                    \n",
    "    def check_exon_validity(self):\n",
    "        \"\"\" The transcript's exons are valid if:\n",
    "            1) Exons are in sorted order (ascending)\n",
    "            2) Exon bounds do not exceed transcript start and end\n",
    "            3) Exons are all on the appropriate chromosome\n",
    "            If these conditions are violated, this function raises an error.\n",
    "        \"\"\"\n",
    "        prev = 0\n",
    "        for exon in self.exons:\n",
    "            if exon.chromosome != self.chromosome:\n",
    "                raise ValueError('Invalid exon in transcript ' + \\\n",
    "                      self.identifier + ': wrong chromosome')\n",
    "            if exon.start < self.start or exon.end > self.end:\n",
    "                print(\"self.start: \" + str(self.start))\n",
    "                print(\"self.end: \" + str(self.end))\n",
    "                print(\"exon.start: \" + str(exon.start))\n",
    "                print(\"exon.end: \" + str(exon.end))\n",
    "                raise ValueError('Invalid exon in transcript ' + \\\n",
    "                      self.identifier + ': (' + str(exon.start) + \"-\" + \\\n",
    "                      str(exon.end) + \\\n",
    "                      ') is located beyond start or end of transcript')\n",
    "            if exon.start <= prev:\n",
    "                # This error would indicate a TALON bug rather than user error,\n",
    "                # so we shouldn't see it. \n",
    "                raise ValueError('Exons of transcript ' + \\\n",
    "                      self.identifier + ' are not stored in ascending order.')\n",
    "            prev = exon.end\n",
    "        return\n",
    "\n",
    "    def get_introns(self):\n",
    "        \"\"\"\n",
    "        Computes introns based on the exon list\n",
    "        \"\"\"\n",
    "        exon_coords = self.get_exon_coords()\n",
    "        intron_list = []\n",
    "\n",
    "        i = 1\n",
    "        while (i < len(exon_coords) - 1):\n",
    "            j = i + 1\n",
    "\n",
    "            intron_list.append(exon_coords[i] + 1)\n",
    "            intron_list.append(exon_coords[j] - 1)\n",
    "            i += 2\n",
    "\n",
    "        return intron_list\n",
    "\n",
    "\n",
    "    def print_transcript(self):\n",
    "        \"\"\" Print a string representation of the Transcript. Good for debugging\n",
    "        \"\"\"\n",
    "        transcript_id = self.identifier\n",
    "        if transcript_id == None:\n",
    "            transcript_id = \"Transcript\"\n",
    "\n",
    "        print(\"\\tLocation: \" + self.chromosome + \":\" + str(self.start) + \"-\" + \\\n",
    "              str(self.end) + \"(\" + self.strand + \")\")\n",
    "\n",
    "        # Print exons\n",
    "        print(\"\\tExons: \" + \"\\n\".join([str(x.start) + \"-\" + str(x.end) for x in self.exons]))\n",
    "        return \n",
    "    \n",
    "    def getTranscriptLineByObject(transcriptObj):\n",
    "        transcriptLine=str(transcriptObj.chromosome+\"\\t\"+transcriptObj.annotations.get('source')+\"\\t\"+\"transcript\"+\"\\t\"+str(transcriptObj.start)\\\n",
    "     +\"\\t\"+str(transcriptObj.end)+\"\\t.\\t\"+transcriptObj.strand+\"\\t.\\t\")\n",
    "        for x in transcriptObj.annotations:\n",
    "            transcriptLine=transcriptLine+str(x+\" \\\"\"+transcriptObj.annotations[x]+\"\\\"; \")\n",
    "        return transcriptLine+\"\\n\"\n",
    "    \n",
    "\n",
    "    def get_transcript_from_gtf(transcript_info):\n",
    "        \"\"\" Uses information from a GTF-formatted transcript entry to create a\n",
    "        Transcript object.\n",
    "            Args:\n",
    "                transcript_info: A list containing fields from a GTF file gene \n",
    "                entry. Example:\n",
    "\n",
    "                chr1\tHAVANA\ttranscript\t12010\t13670\t.\t+\n",
    "                .\tgene_id \"ENSG00000223972.5\"; transcript_id \"ENST00000450305.2\"; \n",
    "                gene_type \"transcribed_unprocessed_pseudogene\"; \n",
    "                gene_status \"KNOWN\"; gene_name \"DDX11L1\"; \n",
    "                transcript_type \"transcribed_unprocessed_pseudogene\"; \n",
    "                transcript_status \"KNOWN\"; transcript_name \"DDX11L1-001\"; \n",
    "                level 2; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; \n",
    "                transcript_support_level \"NA\"; havana_gene \"OTTHUMG00000000961.2\"; \n",
    "                havana_transcript \"OTTHUMT00000002844.2\";\n",
    "        \"\"\"\n",
    "        chromosome = transcript_info[0]\n",
    "        start = int(transcript_info[3])\n",
    "        end = int(transcript_info[4])\n",
    "        strand = transcript_info[6]\n",
    "\n",
    "        if \"transcript_id\" not in transcript_info[-1]:\n",
    "                raise ValueError('GTF entry lacks a transcript_id field')\n",
    "        annotations = extract_transcript_annotations_from_GTF(transcript_info)\n",
    "\n",
    "\n",
    "        gene_id = annotations['gene_id']\n",
    "        transcript_id = annotations['transcript_id']\n",
    "\n",
    "        transcript = Transcript(transcript_id, chromosome, start, end, strand, \n",
    "                                gene_id, annotations)\n",
    "\n",
    "        return transcript\n",
    "\n",
    "def extract_transcript_annotations_from_GTF(tab_fields):\n",
    "    \"\"\"Extracts key-value annotations from the GTF description field\"\"\"\n",
    "\n",
    "    attributes = {}\n",
    "\n",
    "    # remove trailing newline and split by semicolon\n",
    "    description = tab_fields[-1].strip(\"\\n\")\n",
    "    description = description.split(\";\")\n",
    "\n",
    "    # Parse description\n",
    "    for fields in description:\n",
    "        if fields == \"\" or fields == \" \":\n",
    "            continue\n",
    "        fields = fields.split()\n",
    "        if fields[0] == \"\":\n",
    "            fields = fields[1:]\n",
    "\n",
    "        key = fields[0].replace('\"', \"\")\n",
    "        val = \" \".join(fields[1:]).replace('\"', \"\")\n",
    "\n",
    "        attributes[key] = val\n",
    "\n",
    "    # Put in placeholders for important attributes (such as gene_id) if they\n",
    "    # are absent\n",
    "    if \"gene_id\" not in attributes:\n",
    "        attributes[\"gene_id\"] = \"NULL\"\n",
    "\n",
    "    attributes[\"source\"] = tab_fields[1]\n",
    "\n",
    "    return attributes\n",
    " \n",
    "\n",
    "\n",
    "    def get_transcript_from_exon(exon, gene_id, transcript_id):\n",
    "        \"\"\" In rare cases, GTF exons are listed with gene and transcript IDs that\n",
    "            do not have corresponding entries. In this case, we create a transcript\n",
    "            for this exon for bookkeeping purposes.\"\"\"\n",
    "\n",
    "        name = transcript_id\n",
    "        chromosome = exon.chromosome\n",
    "        start = exon.start\n",
    "        end = exon.end\n",
    "        strand = exon.strand\n",
    "        transcript = Transcript(transcript_id, name, None, chromosome, start, end,\n",
    "                                strand, gene_id)\n",
    "        return transcript\n",
    "\n",
    "    def create_novel_transcript(chromosome, start, end, strand, gene_id, counter,\n",
    "                                 exons, introns):\n",
    "        \"\"\" Creates a novel transcript with a unique identifier (obtained using\n",
    "            counter). Returns the transcript object as well as the updated counter.\n",
    "        \"\"\"\n",
    "        counter[\"transcripts\"] += 1\n",
    "        transcript_id = str(counter[\"transcripts\"])\n",
    "\n",
    "        transcript = Transcript(transcript_id, chromosome, start, end, strand, \n",
    "                                gene_id, None)\n",
    "\n",
    "        for exon in exons:\n",
    "            transcript.add_exon(exon)\n",
    "        for intron in introns:\n",
    "            transcript.add_intron(intron)\n",
    "\n",
    "        return transcript\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Gene(object):\n",
    "    \"\"\" Contains high-level information about a gene, such as its identifiers, \n",
    "        genomic location, and transcripts. Does not contain exon information.\n",
    "        Attributes:\n",
    "            - identifier: Accession ID of gene, i.e. an Ensembl ID. Required.\n",
    "            - name: Human-readable name of the gene. This attribute can be left \n",
    "              empty if the gene does not have an assigned name.\n",
    "            - chromosome: Chromosome that the gene is located on (format \"chr1\")\n",
    "            - start: The start position of the gene with respect to the forward \n",
    "              strand (int). Should always be less than or equal to end.\n",
    "            - end: The end position of the gene with respect to the forward strand \n",
    "              (int). Should always be greater than or equal to start.\n",
    "            - strand: \"+\" if the gene is on the forward strand, \"-\" if it is on \n",
    "              the reverse strand\n",
    "            - annotations: a dictionary of miscellaneous annotation categories\n",
    "              extracted from a GTF\n",
    "            \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, identifier, chromosome, start, end, strand, annotations):\n",
    "        start = int(start)\n",
    "        end = int(end)\n",
    "\n",
    "        self.identifier = str(identifier)\n",
    "        self.chromosome = chromosome\n",
    "        self.start = int(start)\n",
    "        self.end = int(end)\n",
    "        self.strand = strand\n",
    "        self.transcripts = {}\n",
    "        self.length = end - start + 1\n",
    "        self.annotations = annotations\n",
    "\n",
    "        if start > end:\n",
    "            raise ValueError(\"\"\"Plus strand gene start must be less than or \n",
    "                             equal to end.\"\"\")\n",
    "\n",
    "    def set_name(self, name):\n",
    "        \"\"\" Sets the name attribute of the Gene to the provided value.\n",
    "        \"\"\"\n",
    "        self.annotations['name'] = name\n",
    "        return\n",
    "\n",
    "    def add_transcript(self, transcript):\n",
    "        \"\"\" Adds a key-value pair (transcript identifier -> Transcript oject)\n",
    "            to the gene's transcript dictionary\n",
    "            Args:\n",
    "                transcript: object of type Transcript. Must overlap with the \n",
    "                location of the gene.\n",
    "        \"\"\"\n",
    "        if transcript.start >= self.end or transcript.end <= self.start:\n",
    "            raise ValueError('Transcript must overlap the gene it is assigned to')\n",
    " \n",
    "        if transcript.gene_id == self.identifier:\n",
    "            # In order to belong to a gene, the transcript gene_id must match\n",
    "            transcript_id = transcript.identifier\n",
    "            self.transcripts[transcript_id] = transcript\n",
    "        else:\n",
    "            raise ValueError('Gene ID of transcript must match gene ' + \\\n",
    "                  'in order for assignment to be made.')\n",
    "        return             \n",
    "\n",
    "\n",
    "    def print_gene(self):\n",
    "        \"\"\" Print a string representation of the Gene. Good for debugging. \"\"\"\n",
    "\n",
    "        if \"name\" in self.annotations != \"\":\n",
    "            # Include name in output if there is one\n",
    "            print(self.identifier + \" (\" + self.annotations['name']  + \"):\")\n",
    "        else:\n",
    "            print(self.identifier + \":\")\n",
    "\n",
    "        print(\"\\tLocation: \" + self.chromosome + \":\" + str(self.start) + \"-\" + \\\n",
    "              str(self.end) + \"(\" + self.strand + \")\")\n",
    "        \n",
    "        # Print transcripts in shorthand \n",
    "        for transcript in self.transcripts:\n",
    "            print(\"\\t Transcript: \" + transcript)\n",
    "\n",
    "        return\n",
    "   \n",
    "    \n",
    "    def getGeneLineByObject(gene):\n",
    "        geneLine=str(gene.chromosome+\"\\t\"+gene.annotations.get('source')+\"\\t\"+\"gene\"+\"\\t\"+str(gene.start)\\\n",
    "         +\"\\t\"+str(gene.end)+\"\\t.\\t\"+gene.strand+\"\\t.\\t\")\n",
    "        for x in gene.annotations:\n",
    "            geneLine=geneLine+str(x+\" \\\"\"+gene.annotations[x]+\"\\\"; \")\n",
    "        return geneLine+\"\\n\"\n",
    "\n",
    "    def get_gene_from_gtf(gene_info):\n",
    "        \"\"\" Creates a Gene object from a GTF file entry\n",
    "            Args:\n",
    "                gene_info: A list containing fields from a GTF file gene entry.\n",
    "                Example:\n",
    "                ['chr1', 'HAVANA', 'gene', '11869', '14409', '.', '+', '.',\n",
    "                'gene_id \"ENSG00000223972.5\";\n",
    "                gene_type \"transcribed_unprocessed_pseudogene\";\n",
    "                gene_status \"KNOWN\"; gene_name \"DDX11L1\"; level 2;\n",
    "                havana_gene \"OTTHUMG00000000961.2\";']\n",
    "        \"\"\"\n",
    "        chromosome = gene_info[0]\n",
    "        start = int(gene_info[3])\n",
    "        end = int(gene_info[4])\n",
    "        strand = gene_info[6]\n",
    "        annotations = extract_gene_annotations_from_GTF(gene_info)\n",
    "        if \"gene_id\" not in gene_info[-1]:\n",
    "                raise ValueError('GTF entry lacks a gene_id field')\n",
    "        gene_id = annotations['gene_id']\n",
    "\n",
    "        gene = Gene(gene_id, chromosome, start, end, strand, annotations)\n",
    "        return gene\n",
    "\n",
    "def extract_gene_annotations_from_GTF(tab_fields):\n",
    "    \"\"\"Parses the description field of a gene GTF in order to organize the \n",
    "       information therein into a dictionary.\n",
    "    \"\"\"\n",
    "\n",
    "    attributes = {}\n",
    "\n",
    "    # remove trailing newline and split by semicolon\n",
    "    description = tab_fields[-1].strip('\\n')\n",
    "    description = description.split(';')\n",
    "\n",
    "    # Parse description\n",
    "    for fields in description:\n",
    "        if fields == \"\" or fields == \" \": continue\n",
    "        fields = fields.split()\n",
    "        if fields[0] == '': fields = fields[1:]\n",
    "\n",
    "        key = fields[0].replace('\"', '')\n",
    "        val = ' '.join(fields[1:]).replace('\"', '')\n",
    "\n",
    "        attributes[key] = val\n",
    "\n",
    "    attributes[\"source\"] = tab_fields[1]\n",
    "\n",
    "    return attributes  \n",
    "\n",
    "    def get_gene_from_exon(exon, gene_id):\n",
    "        \"\"\" In rare cases, GTF exons are listed with gene and transcript IDs that\n",
    "            do not have corresponding entries. In this case, we create a gene\n",
    "            for this exon for bookkeeping purposes.\"\"\"\n",
    "\n",
    "        gene_name = gene_id\n",
    "        chromosome = exon.chromosome\n",
    "        start = exon.start\n",
    "        end = exon.end\n",
    "        strand = exon.strand\n",
    "        gene = Gene(gene_id, gene_name, None, chromosome, start, end, strand)\n",
    "        return gene\n",
    "\n",
    "    def create_novel_gene(chromosome, start, end, strand, counter):\n",
    "        \"\"\" Creates a novel gene with a unique identifier (obtained using\n",
    "            counter). Returns the gene object as well as the updated counter.\n",
    "        \"\"\"\n",
    "        gene_id = str(counter[\"genes\"] + 1)\n",
    "        counter[\"genes\"] += 1\n",
    "        gene = Gene(gene_id, chromosome, start, end, strand, None)\n",
    "        return gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gtf_file(gtf_file):\n",
    "    \"\"\" Reads gene, transcript, and edge information from a GTF file.\n",
    "        Args:\n",
    "            gtf_file: Path to the GTF file\n",
    "        Returns:\n",
    "            genes: A dictionary mapping gene IDs to corresponding gene objects\n",
    "            transcripts: A dictionary mapping gene IDs to corresponding\n",
    "                   transcript objects\n",
    "            exons: A dictionary mapping exon IDs to corresponding edge objects\n",
    "    \"\"\"\n",
    "    genes = {}\n",
    "    transcripts = {}\n",
    "    exons = {}\n",
    "\n",
    "    with open(gtf_file) as gtf:\n",
    "        for line in gtf:\n",
    "            line = line.strip()\n",
    "\n",
    "            # Ignore header\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "\n",
    "            # Split into constitutive fields on tab\n",
    "            tab_fields = line.split(\"\\t\")\n",
    "            chrom = tab_fields[0]\n",
    "            entry_type = tab_fields[2]\n",
    "\n",
    "            # Entry is a gene\n",
    "            if entry_type == \"gene\":\n",
    "                gene = Gene.get_gene_from_gtf(tab_fields)\n",
    "                native_id = gene.identifier\n",
    "                genes[native_id] = gene\n",
    "\n",
    "            # Entry is a transcript\n",
    "            elif entry_type == \"transcript\":\n",
    "                transcript = Transcript.get_transcript_from_gtf(tab_fields)\n",
    "                gene_id = transcript.gene_id\n",
    "                if gene_id in genes:\n",
    "                    genes[gene_id].add_transcript(transcript)\n",
    "                native_id = transcript.identifier\n",
    "                transcripts[native_id] = transcript\n",
    "\n",
    "            # Entry is an edge\n",
    "            elif entry_type == \"exon\":\n",
    "                exon = Edge.create_edge_from_gtf(tab_fields)\n",
    "                # This ID is used because of a rare GENCODE bug\n",
    "                location_exon_id = exon.identifier\n",
    "                exons[location_exon_id] = exon\n",
    "\n",
    "                transcript_id = list(exon.transcript_ids)[0]\n",
    "                gene_id = exon.annotations[\"gene_id\"]\n",
    "\n",
    "                if location_exon_id not in exons:\n",
    "                    # Add the new edge to the data structure\n",
    "                    exons[location_exon_id] = exon\n",
    "                else:\n",
    "                    # Update existing exon entry, including its transcript set\n",
    "                    exon = exons[location_exon_id]\n",
    "                    exon.transcript_ids.add(transcript_id)\n",
    "\n",
    "                if transcript_id in transcripts:\n",
    "                    currTranscript = transcripts[transcript_id]\n",
    "                    currTranscript.add_exon(exon)\n",
    "\n",
    "    return genes, transcripts, exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def printNumberOfGeneTranscriptsExons(genes, transcripts, exons):\n",
    "    print('geneCount',len(genes))\n",
    "    print('transcriptCount',len(transcripts))\n",
    "    print('exonCount',len(exons))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################################################################\n",
    "def getGeneWithTranscriptsAndExonsAsLines(gene):\n",
    "    geneLine=Gene.getGeneLineByObject(gene)\n",
    "    geneWithTranscriptsAndExonsLines=geneLine\n",
    "    #geneWithTranscriptAndExons=geneLine\n",
    "    for transcriptID in gene.transcripts:\n",
    "        transcriptObj=gene.transcripts.get(transcriptID)\n",
    "        transcriptLine=Transcript.getTranscriptLineByObject(transcriptObj)\n",
    "        geneWithTranscriptsAndExonsLines=geneWithTranscriptsAndExonsLines+transcriptLine\n",
    "        for exonObj in transcriptObj.exons:\n",
    "            exonLine=Edge.getExonLineByObject(exonObj)\n",
    "            geneWithTranscriptsAndExonsLines=geneWithTranscriptsAndExonsLines+exonLine\n",
    "    return geneWithTranscriptsAndExonsLines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################################################################    \n",
    "def writeGeneToTheFile(singleGene,gtfFilePath):\n",
    "    gtfFileObj = open(gtfFilePath, \"a\")\n",
    "    for i in range(len(singleGene)):\n",
    "        gtfFileObj.write(singleGene[i])\n",
    "    gtfFileObj.close()\n",
    "    \n",
    "#########################################################################################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validateGene(gene):\n",
    "    if len(gene.transcripts)==0:\n",
    "        raise ValueError(\"Gene : \\t\",gene.identifier, \"\\t has no Transcripts\")\n",
    "    if gene.start > gene.end:\n",
    "        raise ValueError(\"Gene : \\t\",gene.identifier ,\"\\t start > end\")\n",
    "    for transcriptID in list(gene.transcripts):\n",
    "        transcriptObj=gene.transcripts.get(transcriptID)\n",
    "        if len(transcriptObj.exons)==0:\n",
    "            raise ValueError(\"Transcript : \\t\",transcriptObj.identifier ,\"\\t has no exons\")\n",
    "        if transcriptObj.start > transcriptObj.end:\n",
    "            raise ValueError(\"Transcript : \\t\",transcriptObj.identifier, \"\\t start > end\")\n",
    "        for exon in transcriptObj.exons:\n",
    "            if exon.start>exon.end:\n",
    "                raise ValueError(\"exon : \\t\",exon.identifier ,\"\\t start > end\")\n",
    "            if exon.start==0 or exon.end==0:\n",
    "                raise ValueError(\"exon : \\t\",exon.identifier, \"\\t start or end is zero\")\n",
    "    return True\n",
    "\n",
    "\n",
    "def validateAndWriteGTFToFile(CleanedSameStrandGenes,cleanedForSenseAndAntisenseFile):\n",
    "    if os.path.exists(cleanedForSenseAndAntisenseFile):\n",
    "        os.remove(cleanedForSenseAndAntisenseFile)\n",
    "    for geneID in CleanedSameStrandGenes.keys():\n",
    "        gene=CleanedSameStrandGenes.get(geneID)\n",
    "        if validateGene(gene)==True:\n",
    "            geneWithTranscriptsAndExonsLines=getGeneWithTranscriptsAndExonsAsLines(gene)\n",
    "            writeGeneToTheFile(geneWithTranscriptsAndExonsLines,cleanedForSenseAndAntisenseFile)\n",
    "        else:\n",
    "            raise ValueError(\"Gene : \\t\",gene.identifier, \"\\t has some problems\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
